### ISSUE TITLE
Segmentation fault when running KAT

### ISSUE THREAD

**Issue by zhiliangc** [TGAC/KAT#31]
I I compiled with 
```
1) boost/1.58      2) libtool/2.4.6   3) gnuplot/5.0.1   4)  gcc/5.1.0 
```
and finally manage to install KAT V2.0.8 but still couldn't run it :(
I need to run it on a HPC so I've written a job script and submitted it to the compute nodes. 
However, it gave me the error message
```
~/KAT/KAT-2.0.8/bin/kat comp -C -D --d1_bins 2000 *_L001_R?.fastq 
Kmer Analysis Toolkit (KAT) V2.0.8

Running KAT in COMP mode
------------------------

Segmentation fault
```
I tried to run it on head node without submitting to compute nodes, it's still the same. 

Our HPC runs on SUSE Linux Enterprise Server 11 (x86_64) and PBSPro_11.2.0.113417

**Comment by zhiliangc** [comment]
btw, after installation, ```jellyfish``` was named as ```jellyfishk ``` under bin for some reason, I have to rename it to ```jellyfish```

**Comment by zhiliangc** [comment]
Just realised if I've taken out ```--d1_bins```, it works perfectly

**Comment by maplesond** [comment]
Great that you've got it working!  I'll take a look and figure out why d1_bins is causing an issue.

The change of jellyfish binary is intentional to avoid clashing with any existing jellyfish installation... see https://github.com/TGAC/KAT/issues/29  You should only need to use it if you want to run `jellyfish count` independanyly from KAT though, so you shouldn't need to rename it.

**Comment by maplesond** [comment]
Also I noticed your comp command is a bit ambiguous, you use *_L001_R?.fastq to define your input groups.  This will work ok if this pattern matches 2 files (compare first file against the second), but will produce variable results if there are any other number of files matching the pattern.  It will fail with 1, or more than 3.  With 3 it will probably not do what you expect.  In other KAT tools this pattern would essentially cat the files matching the pattern together.  You can replicate that behaviour in KAT by surrounding the pattern in single quotes.

**Comment by zhiliangc** [comment]
The reason I rename it was it actually gave me an error like this:
```
Kmer Analysis Toolkit (KAT) V2.0.8

inc/kat_fs.hpp(129): Throw in function kat::KatFS::KatFS(const char*)
Dynamic exception type: boost::exception_detail::clone_impl<kat::FileSystemException>
std::exception::what: std::exception
[kat::FileSystemError*] = Could not find jellyfish executable at: /home/zxc561/KAT/KAT-2.0.8/bin/jellyfish
```
I guess there must be somewhere in the code still trying to run ```jellyfish``` instead of ```jellyfishk``` ?
And thanks for pointing out the problem with my command, there're only two fastq files in that folder so I was just being a bit lazy of typing all the file names.

**Comment by zhiliangc** [comment]
Tried to run
```
~/KAT/2.0.8/bin/kat comp -t 16 -C -D -o reads_vs_abyss_k27 'CS3270_A8733_GCCAAT_L001_R
?.fastq' ../abyss/fusarium_k27-contigs.fa
```
and it gave me ```77183 Segmentation fault``` again.

I was wondering whether it's something related to our HPC as it was exactly the same command  @bjclavijo ran when he came and help with our workshop last week (using ```kat/2.0.6```)

**Comment by tseemann** [comment]
@zhiliangc Did you mean to use single quites with the `?` wildcard with R1/R2 ?  The `?` won't expand unless you use double quotes.

**Comment by maplesond** [comment]
@zhiliangc That's odd.  Clearly we shouldn't get a segmentation fault here, and at the very least get a suitable error message.  I'm not sure what's causing the problem, do you get any output at all except for a segmentation fault?

@tseemann Single quotes work fine on my system, although double quotes work ok too: 

dan@dan-home-mint ~/temp/kat $ ~/dev/KAT/src/kat comp -t 2 'DRR015910_?.1M.fastq' ecoli_k12.fasta 
Kmer Analysis Toolkit (KAT) V2.0.8

Running KAT in COMP mode
------------------------

Input is a sequence file.  Counting kmers for DRR015910_1.1M.fastq DRR015910_2.1M.fastq ... done.  Time taken: 14.7s

Input is a sequence file.  Counting kmers for ecoli_k12.fasta ... done.  Time taken: 0.8s

Comparing hashes ... done.  Time taken: 8.7s

Merging results ... done.  Time taken: 0.4s

Saving results to disk ... done.  Time taken: 0.2s

Creating plot ... done.  Time taken: 1.1s

....

It's useful to use single quotes when working on scheduling systems (particularly when not using scripts) to avoid clashes with the double quotes that wrap the command.  Also, what's currently preventing you from building KAT?

**Comment by tseemann** [comment]
@maplesond I can now build it from HEAD but I was waiting for a Release to update Brew.

**Comment by zhiliangc** [comment]
@maplesond Here's everything I got
```
Kmer Analysis Toolkit (KAT) V2.0.8

Running KAT in COMP mode
------------------------

/var/spool/PBS/mom_priv/jobs/1248199.admin.SC: line 22: 77183 Segmentation fault      ~/KAT/2.0.8/bin/kat comp -t 16 -C -D -o reads_vs_abyss_k27 'CS3270_A8733_GCCAAT_L001_R
?.fastq' ../abyss/fusarium_k27-contigs.fa
```

It jumped striaght into Segmentation fault

**Comment by zhiliangc** [comment]
I've tried several commands.
It works with both  ```-o``` and ```--d1_bins ``` options, e.g.
```
~/KAT/2.0.8/bin/kat comp -t 16 -C -D -o reads_vs_abyss_k27 --d1_bins 2000 "CS3270_A8733_GCCAAT_L001_R?.fastq" ../abyss/fusarium_k27-contigs.fa 
```
If I omitted the value for ```-o```,e.g.
```
kat comp -C -D --d1_bins 2000 CS3270_A8733_GCCAAT_L001_R1.fastq CS3270_A8733_GCCAAT_L001_R2.fastq
```
It outputs the segmentation fault.

**Comment by maplesond** [comment]
Ok, thanks for narrowing this down.  I'll try and find the root cause of the problem and issue a fix in the week.

**Comment by maplesond** [comment]
@tseemann Apologies for not uploading a release earlier.  I'll do this now, although I'll probably also make a V2.0.9 release to fix any critical bugs.

**Comment by maplesond** [comment]
@zhiliangc I can't reproduce this one on my system, with any combination of included or excluded parameters.  If you can work around for the time being I suggest you carry on with that.  If you can reproduce in V2.1 when that's released, or if you have any further indications of how to reproduce it I'll take another look.

**Comment by zhiliangc** [comment]
Thank you

**Comment by maplesond** [comment]
I can no reproduce this one, or something similar.  It seems sometimes I get a error after counting a file using jellyfish:

```
Kmer Analysis Toolkit (KAT) V2.1.0

Running KAT in COMP mode
------------------------

Input is a sequence file.  Counting kmers for ccoli.R1.100K.fastq...*** Error in `/home/maplesod/dev/KAT/src/kat': free(): invalid next size (fast): 0x0000000000780c60 ***
```

Slightly modifying the command line can make the problem go away (doesn't seem to matter what happens to the command line, e.g. extra options, different length arguments, etc).  Anyway, I have a case which produces the problem, so I'll look into this and see if there's some way of fixing it.

**Comment by maplesond** [comment]
I can't replicate this issue in the latest version of the codebase, so hopefully this will be fixed in V2.1.

**Comment by maplesond** [comment]
Should be fixed in the latest release: https://github.com/TGAC/KAT/releases/tag/Release-2.1.0.  Please reopen if there are any issues.
